### PR TITLE
ISPN-13725 Integration with OpenTelemetry Tracing

### PR DESCRIPTION
**Pull request by fax4ever**
https://issues.redhat.com/browse/ISPN-13725

now (re)based on #10196 (HotRod v4 complete PR).

### CODE REVIEW

**Comment by fax4ever** [comment]
RequestTracingIT fails since we need to add more dependencies on the server

**Comment by fax4ever** [comment]
Jenkins retest this please

**Review by pruivo — CHANGES_REQUESTED** [review]

**Comment by fax4ever** [comment]
Thanks for the review @pruivo. I applied the changes and I also use `CONTAINER` as a run mode in the test for the tracing.
That is was I thought I was using :( How silly of me

File: pom.xml [hunk 1]
```diff
@@ -407,12 +407,10 @@
       <versionx.org.springframework.spring-expression>${version.spring5}</versionx.org.springframework.spring-expression>
       <versionx.org.springframework.spring-jdbc>${version.spring5}</versionx.org.springframework.spring-jdbc>
       <versionx.org.springframework.spring-test>${version.spring5}</versionx.org.springframework.spring-test>
-      <!-- Version 0.33.0 is not compatible -->
-      <versionx.org.opentracing.api>0.31.0</versionx.org.opentracing.api>
-      <versionx.io.jaegertracing.jaeger-client>1.6.0</versionx.io.jaegertracing.jaeger-client>
       <versionx.org.testcontainers.testcontainers>1.17.2</versionx.org.testcontainers.testcontainers>
       <versionx.io.jaegertracing.jaeger-testcontainers>0.7.0</versionx.io.jaegertracing.jaeger-testcontainers>
       <versionx.micrometer>${version.micrometer}</versionx.micrometer>
+      <versionx.opentelemetry>${version.opentelemetry}</versionx.opentelemetry>
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply pom.xml]: thanks applied

File: build/configuration/pom.xml [hunk 2]
```diff
@@ -199,6 +199,9 @@
       <version.hdrhistogram>2.1.12</version.hdrhistogram>
       <version.latencyutils>2.0.2</version.latencyutils>
 
+      <version.opentelemetry>1.15.0</version.opentelemetry>
+      <version.opentelemetry.sdk-autoconfigure>${version.opentelemetry}-alpha</version.opentelemetry.sdk-autoconfigure>
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply build/configuration/pom.xml]: thanks applied

File: server/core/src/main/java/org/infinispan/server/core/RequestTracer.java [hunk 3]
```diff
@@ -1,41 +1,36 @@
...
-               factoryInstance = tracerFactoryClass.getDeclaredConstructor().newInstance();
-            }
-            adapter = new TracerAdapter((Tracer) getTracer.invoke(factoryInstance));
-            log.infof("OpenTracing implementation loaded: %s", adapter);
+            OpenTelemetry openTelemetry = AutoConfiguredOpenTelemetrySdk.builder()
+                  .build()
+                  .getOpenTelemetrySdk();
+
+            tracer = openTelemetry.getTracer("org.infinispan.tracing", "1.0.0");
+            log.infof("OpenTelemetry tracer loaded: %s", tracer);
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply server/core/src/main/java/org/infinispan/server/core/RequestTracer.java:28]: right, thanks!

**Comment by fax4ever** [comment]
Fixed. Thanks @pruivo

**Comment by fax4ever** [comment]
With the last two commits:
1. I changed the way to produce and use the OpenTelemetry instance, the Infinispan D.I. is used now to produce and lookup the instance. This will give us more flexibility.
2. I started to work on context propagation, I created a test to start to work on it.

File: client/hotrod-client/pom.xml [hunk 4]
```diff
@@ -177,6 +177,17 @@
          <artifactId>jakarta.transaction-api</artifactId>
       </dependency>
 
+      <dependency>
+         <groupId>io.opentelemetry</groupId>
+         <artifactId>opentelemetry-api</artifactId>
+      </dependency>
```
Comment (tristantarrant) [inline client/hotrod-client/pom.xml:183]: While for the server, adding a hard dependency on the OTel API is ok-ish (although I'd still prefer some kind of lightweight wrapper), it should be absolutely optional for the client.

Comment (fax4ever) [reply client/hotrod-client/pom.xml:183]: OK let's introduce a lightweight wrapper, to address the optionality and to reduce (remove) the overall dependency from the OpenTelemetry API

File: server/core/src/main/java/org/infinispan/server/core/RequestTracer.java [hunk 5]
```diff
@@ -44,54 +39,20 @@ public static void start() {
...
-         adapter.stop();
-      }
+      // theoretically OpenTelemetry should stop by itself (IIC)
+      GlobalOpenTelemetry.resetForTest();
    }
 
-   public static Object requestStart(String operationName) {
-      if (adapter != null) {
-         return adapter.requestStart(operationName);
+   public static Span requestStart(String operationName) {
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply server/core/src/main/java/org/infinispan/server/core/RequestTracer.java]: @pruivo I made the change, I added a class to keep the abstraction

File: client/hotrod-client/pom.xml (hunk 4 above)
Comment (fax4ever) [reply client/hotrod-client/pom.xml:210]: done with last (forced) push! thanks for the suggestion

**Comment by fax4ever** [comment]
@pruivo @tristantarrant with the last change I added some abstractions to handle the optionality of the dependency from the client and to reduce (remove) the overall dependency from the OpenTelemetry API o the server

**Comment by fax4ever** [comment]
CI looks good!

**Comment by tristantarrant** [comment]
There's overlap between this and https://github.com/infinispan/infinispan/pull/10196

**Comment by tristantarrant** [comment]
I would separate the Hot Rod stuff in a separate PR

**Comment by fax4ever** [comment]
> I would separate the Hot Rod stuff in a separate PR

Done. I extracted #10202 as a base for this contribution.

**Comment by fax4ever** [comment]
I updated the contribution to be based on the changes on #10202.
That is the fact that the `Map<String, String>` header params is now a `Map<String, byte[]>`

**Comment by fax4ever** [comment]
Rebased on current master

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/HeaderParams.java [hunk 6]
```diff
@@ -23,7 +24,7 @@ public class HeaderParams {
    final long messageId;
    int topologyAge;
    DataFormat dataFormat;
-   Map<String, byte[]> otherParams;
+   Map<String, byte[]> otherParams = new HashMap<>();
```
Comment (tristantarrant) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/HeaderParams.java:27]: I would avoid allocation unless header params are present

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/PutStreamOperation.java [hunk 7]
```diff
@@ -45,7 +45,7 @@ public PutStreamOperation(Codec codec, ChannelFactory channelFactory,
                              long lifespan, TimeUnit lifespanTimeUnit, long maxIdle, TimeUnit maxIdleTimeUnit,
                              ClientStatistics clientStatistics) {
       super(PUT_STREAM_REQUEST, PUT_STREAM_RESPONSE, codec, channelFactory, key, keyBytes, cacheName, topologyId,
-            flags, cfg, null, clientStatistics);
+            flags, cfg, null, clientStatistics, null);
```
Comment (tristantarrant) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/PutStreamOperation.java:48]: Why are you not passing the telemetry service here ? This operation is functionally equivalent to `PutOperation`

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/GetOperation.java [hunk 8]
```diff
@@ -27,7 +27,8 @@
    public GetOperation(Codec codec, ChannelFactory channelFactory,
                        Object key, byte[] keyBytes, byte[] cacheName, AtomicInteger topologyId, int flags,
                        Configuration cfg, DataFormat dataFormat, ClientStatistics clientStatistics) {
-      super(GET_REQUEST, GET_RESPONSE, codec, channelFactory, key, keyBytes, cacheName, topologyId, flags, cfg, dataFormat, clientStatistics);
+      super(GET_REQUEST, GET_RESPONSE, codec, channelFactory, key, keyBytes, cacheName, topologyId, flags, cfg,
+            dataFormat, clientStatistics, null);
```
Comment (tristantarrant) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/GetOperation.java:31]: It seems like you are not adding telemetry to read operations. Why ?

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/logging/Log.java [hunk 9]
```diff
@@ -386,4 +386,9 @@ public interface Log extends BasicLogger {
    @LogMessage(level = INFO)
    @Message(value = "Native IOUring transport not available, using NIO instead: %s", id = 4108)
    void ioUringNotAvailable(String cause);
+
+   @LogMessage(level = DEBUG)
+   @Message(value = "OpenTelemetry API is not present in the classpath. No tracing client context will be propagated.", id = 4109)
```
Comment (tristantarrant) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/logging/Log.java:391]: "Client context tracing will not be propagated"

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/telemetry/impl/TelemetryService.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
+package org.infinispan.client.hotrod.telemetry.impl;
+
+import java.nio.charset.StandardCharsets;
+
+import org.infinispan.client.hotrod.impl.protocol.HeaderParams;
+
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Context;
+
+public class TelemetryService {
```
Comment (tristantarrant) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/telemetry/impl/TelemetryService.java:10]: Are you sure this won't cause `ClassNotFoundException` if the OTel API is not present on the classpath ?
Maybe turn `TelemetryService` into an interface and have a single implementing class `OpenTelemetryService`.

**Review by tristantarrant — CHANGES_REQUESTED** [review]
I have made comments on the legacy client, but they apply to the new one as well.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/GetOperation.java (hunk 8 above)
Comment (fax4ever) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/GetOperation.java:31]: In this first version of tracing implementation I would avoid to trace all read operations. In the next version I would add a tracing filter to allow the user to set up what to trace and what not.

**Review by tristantarrant — APPROVED** [review]
LGTM. Waiting for CI

**Comment by tristantarrant** [comment]
Merged, thanks

**Comment by fax4ever** [comment]
Thanks Tristan!

**Pull request merged**